homeassistant/components/sensor/pjm.py

```diff
@@ -0,0 +1,135 @@
+}
+
+TYPES_SCHEMA = vol.In(SENSOR_TYPES)
+
+SENSORS_SCHEMA = vol.Schema({
+    vol.Required(CONF_SENSOR_TYPE): TYPES_SCHEMA,
+    vol.Optional(CONF_ZONE): cv.string,
+    vol.Optional(CONF_NAME): cv.string,
+})
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_MONITORED_FEEDS): [SENSORS_SCHEMA],
+})
+
+
+@asyncio.coroutine
+def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+    """Set up the PJM sensor."""
+    websession = async_get_clientsession(hass)
+    dev = []
+
+    for variable in config[CONF_MONITORED_FEEDS]:
+        dev.append(PJMSensor(
+            hass.loop, websession, variable[CONF_SENSOR_TYPE],
+            variable.get(CONF_ZONE), variable.get(CONF_NAME)))
+
+    async_add_devices(dev, True)
+
+
+class PJMSensor(Entity):
+    """Implementation of a PJM sensor."""
+
+    def __init__(self, loop, websession, sensor_type, zone, name):
+        """Initialize the sensor."""
+        self.loop = loop
+        self.websession = websession
+        if name:
+            self._name = name
+        else:
+            self._name = SENSOR_TYPES[sensor_type][0]
+            if sensor_type == CONF_INSTANTANEOUS_ZONE_LOAD:
+                self._name += ' ' + zone
+        self.type = sensor_type
+        self._state = None
+        if zone:
+            self.zone = zone
+        self._unit_of_measurement = SENSOR_TYPES[sensor_type][1]
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def icon(self):
+        """Icon to use in the frontend, if any."""
+        if self.type == CONF_INSTANTANEOUS_ZONE_LOAD:
+            return ICON_POWER
+        if self.type == CONF_INSTANTANEOUS_TOTAL_LOAD:
+            return ICON_POWER
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self._state
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit of measurement of this entity, if any."""
+        return self._unit_of_measurement
+
+    @asyncio.coroutine
+    def async_update(self):
+        """Get the PJM data from the web service."""
+        from bs4 import BeautifulSoup
+
+        try:
+            with async_timeout.timeout(60, loop=self.loop):
+                response = yield from self.websession.get(_RESOURCE)
+                text = yield from response.text()
+                soup = BeautifulSoup(text, "html.parser")
+                search_text = PJM_RTO_TOTAL
+                if self.type == CONF_INSTANTANEOUS_ZONE_LOAD:
+                    search_text = self.zone + ' Zone'
+                # Find the table row that has the zone we want
+                found_text = soup.find('td', text=search_text)
+                # The next cell contains the power data
+                self._state = found_text.find_next_sibling('td').text
+
+        except (asyncio.TimeoutError, aiohttp.ClientError) as err:
```

Comment:
These items convince me: all of the bs4 logic should be in a separate package that you create. Furthermore, that package should produce its own exception types for you to catch (rather than catching generic types or types from various other libraries).

Reply:
It seems like overkill to create a separate package for such a simple sensor that's just scraping data from a few table cells on a web page.  There are other HA sensors that do far more than this one that don't rely on separate packages

Reply:
@balloob?